### ISSUE TITLE
modifications in imported modules are not seen?

### ISSUE THREAD

**Issue by spyder-issues-migrator** [spyder-ide/spyder#241]
_From [asterbom...@gmail.com](https://code.google.com/u/111310574906366460267/) on 2010-06-24T03:50:50Z_

It appears that there might be a problem with importing of modules in the interactive console (spider 1.1.0rc2).

My code has: 

import engineeringNumbers as en  
print en.num2str(33)

This works fine in the External console, but not in the interactive console, this produces:
    print en.num2str(33)
AttributeError: 'module' object has no attribute 'num2str'

After trying various things the num2str function is correctly shown in the class browser (not so at first), but the interactive console still does not find the function.
This function was added on the fly, and it seems that the interactive console is somehow still using an old version of this module.
Deleleting the .pyc version of the module does not help.
Restarting the interactive interpreter also does *not* help.
Closing spider and re-opening it is the only way I found that solves the problem.

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=241_

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2010-06-24T02:04:10Z_

That is the expected behaviour. Python does not reload modules which have already been imported once before (restarting the interactive console's interpreter does not change this because the interactive console is running in the same thread as Spyder's main window: the whole application has to be restarted). 

This being said, a feature has been implemented in Spyder to force Python to deeply reload modules when importing them. This feature is disabled by default to make sure that the user is conscious of this behaviour which is inherent to Python itself (the user may be confronted to this behaviour outside Spyder, so the best is to not hide this limitation by default).

Now that you are aware of this limitation, you may enable the option "Force Python to reload modules" in  the interactive console menu settings.

**Status:** WontFix

**Comment by spyder-issues-migrator** [comment]
_From [asterbom...@gmail.com](https://code.google.com/u/111310574906366460267/) on 2010-06-25T23:20:43Z_

I understand your point.  But IMHO when the user "restarts" the interactive interpreter then the behavior he really expects is that he gets a completely clean and fresh started interpreter!  I'm sure that it is possible for problems to occur that are not helped by "force reload modules", like instabilities in a library or something.  

Perhaps it would be possible to not embed the interactive interpreter in the main application, but instead run it as a separate process (that uses pipe communication or something), so it can be truly killed and restarted?

It might even be useful then to start 2 interactive interpreters, one for immediate use, and one that sits waiting in the background.  Then at the next run, the first is killed and the second is used automatically, while a new one is already started again in the background.  This mode of operation would combine the advantages of the external console (fresh interpreter) with those of the interactive console (no dead time for the user and full access to variables).

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2010-06-26T01:10:45Z_

In Spyder v1.x, the whole point of the interactive console was to be in the same thread/process as the main application, allowing to embed matplotlib figures directly into Spyder's GUI and solving the "Input Hook" issues of PyQt on Windows platforms (i.e. allowing to use matplotlib in interactive mode without blocking the interpreter). If you to run the interpreter in a separate process, you may start as many interpreters as you want in the external console with almost the same features as in the interactive console (except for the two specific features mentioned above: embedding matplotlib figures and input hook issues).

So basically, when you ask if it would be possible not to embed the intepreter in the main application but instead run it in a separate process... well, it's already implemented in the external console!

BTW, in Spyder v2.x, the interactive console has been renamed into "internal console" (it's multithreaded and for internal debugging only -- hidden by default) and the external console (renamed into "Console") is *the* unique console. Why is that? Because thanks to IPython integration, there is no longer "input hook" issues on Windows platforms.
